### PR TITLE
Rewrite, clarify, and add models

### PR DESCRIPTION
**Pull request by MorganEPatch**
Large-scale rewrite and clarification of the API docs. Many parts of the documentation have gone unedited with changes we've agreed upon, and were in several places self-contradictory. Also added model documentation, which shows what fields exist on what objects, where they can be set to empty values, etc.

### CODE REVIEW

File: source/draft_auth.rst [hunk 1]
```diff
@@ -5,15 +5,14 @@ Authorization
 =============
 
 TimeSync allows for various forms of authentication determined by the
-implementation. The general scheme an authentication module follows looks like
-this:
+implementation. The general scheme of an authentication module follows looks
```
Comment (athai) [inline source/draft_auth.rst]: Choose either "follows", "looks like this", or add a conjunction between the two.

File: source/draft_errors.rst [hunk 2]
```diff
@@ -105,22 +107,26 @@ duration field.)
...
 ---------------------
 
-5. Invalid identifier
+5. Invalid Identifier
 ---------------------
 
-This error would be returned when an identifier field (e.g. time ID or activity
+This error would be returned when an identifier field (e.g. time UUID or activity
 slug) is malformed or otherwise not valid for use. This is to be distinguished
 from Object not found: Object not found occurs when a perfectly valid,
```
Comment (athai) [inline source/draft_errors.rst]: Forgot to capitalize "Object not found" twice here.

File: source/draft_model.rst [hunk 3]
```diff
@@ -0,0 +1,76 @@
+.. _draft_model:
+
+================
+API Object Model
+================
+
+Below is a listing of the structure of objects in the timesync API.
```
Comment (athai) [inline source/draft_model.rst]: timesync -> TimeSync

File: source/draft_model.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+Each object is listed, with the fields it consists of. Type represents the type
```
Comment (athai) [inline source/draft_model.rst]: Is the comma necessary here?

File: source/draft_model.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+of data passed; it may represent data types (such as "string" or "array") or
+formats (such as "ISO date"). Required_post represents whether "empty" values
```
Comment (athai) [inline source/draft_model.rst]: Required_post -> POST Required (if only to be consistent with the tables)

File: source/draft_api.rst [hunk 6]
```diff
@@ -717,13 +758,15 @@ In addition, each user has a role within each project to which they belong:
 * data viewer
```
Comment (athai) [inline source/draft_api.rst]: Question: In the schema, 'data viewer' is listed as 'spectator'. Which are we going with?

I can change that bit if you want to go with 'data viewer'.

File: source/draft_errors.rst [hunk 7]
```diff
@@ -165,23 +155,23 @@ server is running.
...
         "status": 401,
         "error": "Authentication failure",
-        "text": "Invalid password" / "Bad oAuth token" / etc
+        "text": "Invalid username or password" / "Bad oAuth token" / etc
     }
 
 ----------------------
 
-8. Slug already exists
+7. Slug already exists
```
Comment (athai) [inline source/draft_errors.rst]: All errors following this one are also not completely capitalized.

File: source/draft_api.rst [hunk 8]
```diff
@@ -309,16 +348,16 @@ For example:
       "slugs":["testing", "test"],
       "updated_at": 2015-04-18,
```
Comment (athai) [inline source/draft_api.rst]: This object has 2 "updated_at" fields

File: source/draft_api.rst [hunk 9]
```diff
@@ -309,16 +348,16 @@ For example:
...
+      "updated_at": "2014-04-18",
       "revision":2,
       "parents":
         [
           {
             "name":"Testing Infrastructure",
             "slugs":["testing", "tests"],
-            "created_at": 2015-04-17,
+            "created_at": "2015-04-17",
             "deleted_at": null,
```
Comment (athai) [inline source/draft_api.rst]: This object has two "deleted_at" fields

File: source/draft_api.rst [hunk 10]
```diff
@@ -328,37 +367,70 @@ For example:
...
+        "deleted_at": null,
+        "updated_at": "2014-04-18",
+        "revision":2,
+        "parents":
+          [
+            {
+              "name":"Testing Infrastructure",
+              "slugs":["testing", "tests"],
+              "created_at": "2015-04-17",
+              "deleted_at": null,
```
Comment (athai) [inline source/draft_api.rst]: This has 2 "deleted_at" fields

File: source/draft_api.rst [hunk 11]
```diff
@@ -328,37 +367,70 @@ For example:
...
+        "deleted_at": null,
+        "updated_at": "2014-04-23",
+        "revision":2,
+        "parents":
+          [
+            {
+              "name":"Testing Infrastructure",
+              "slugs":["testing", "tests"],
+              "created_at": "2015-04-17",
+              "deleted_at": null,
```
Comment (athai) [inline source/draft_api.rst]: This has 2 "deleted_at" fields

File: source/draft_api.rst [hunk 12]
```diff
@@ -714,16 +751,18 @@ managers/viewers of all projects.
...
+or may be a member and manager but not viewer), and are defined by their
+permissions:
 
 * a member has permission to write time entries
-* a data viewer may view time entries
+* a spectator may view time entries
 * a project manager may update the project information
 
-A user may be a member, viewer, or manager of multiple projects, and a project
+A user may be a member, viewer, and/or manager of multiple projects, and a project
```
Comment (athai) [inline source/draft_api.rst]: viewer -> spectator

File: source/draft_api.rst [hunk 13]
```diff
@@ -714,16 +751,18 @@ managers/viewers of all projects.
...
 may have multiple members, viewers, and managers.
```
Comment (athai) [inline source/draft_api.rst]: viewers -> spectators

File: source/draft_api.rst [hunk 14]
```diff
@@ -714,16 +751,18 @@ managers/viewers of all projects.
 In addition, each user has a role within each project to which they belong:
 
 * member
-* data viewer
+* spectator
 * project manager
 
-These roles exist independently, and are defined by their permissions:
+These roles exist independently (for example, a user may be only a spectator,
+or may be a member and manager but not viewer), and are defined by their
```
Comment (athai) [inline source/draft_api.rst]: viewer -> spectator

File: source/draft_model.rst [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+------
+
+Common
+------
+
+These fields are shared between all objects. Do not pass any of them in a POST
+request; they are provided automatically.
+
+==========  ===============   =============  ======================================  ===========================================
```
Comment (athai) [inline source/draft_model.rst]: Your tables don't exist. 

https://github.com/osuosl/timesync/blob/tristan/rewrite-add-models/source/draft_model.rst

File: source/draft_model.rst [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+deleted_at  ISO date or null  false          The date this revision was deleted      Always null unless ?include_deleted is used
+parents     array             false          The previous revisions of this object   Only provided if ?revisions=true is used
+==========  ================  =============  ======================================  ===========================================
+
+-----
+
+Times
+-----
+
+===========  ===============   =============  ======================================  ============================================
```
Comment (athai) [inline source/draft_model.rst]: The 'Times' and 'Activities' tables still aren't being generated.

https://github.com/osuosl/timesync/blob/tristan/rewrite-add-models/source/draft_model.rst#times

**Comment by athai** [comment]
Looks good! +1

File: source/draft_errors.rst [hunk 17]
```diff
@@ -165,23 +155,23 @@ server is running.
...
 
-This error is returned when a new object is being created but the slugs passed
-in contain a slug that already exists.
+This error is returned when a new object is being created but the slugs provided
+contain a slug that already exists.
 
 .. code-block:: javascript
 
     {
         status: 409,
```
Comment (athai) [inline source/draft_errors.rst]: In the code-block examples above, each fieldname is encased in quotes. Starting with Error 7, that consistency is broken (e.g. status compared to "status").

Also, if you could fix this so that the text isn't highlighted red, that'd be ace. 
https://github.com/osuosl/timesync/blob/tristan/rewrite-add-models/source/draft_errors.rst#bad-object

**Comment by pop** [comment]
lgtm. I'll find a middle ground between your changes and mine in #39.
+1

File: source/draft_api.rst [hunk 18]
```diff
@@ -307,58 +346,87 @@ For example:
...
-      "updated_at": 2014-04-18,
+      "updated_at": "2014-04-18",
       "revision":2,
       "parents":
         [
           {
             "name":"Testing Infrastructure",
             "slugs":["testing", "tests"],
-            "created_at": 2015-04-17,
+            "created_at": "2015-04-17",
```
Comment (athai) [inline source/draft_api.rst]: The year here is different than in revision 2.

File: source/draft_api.rst [hunk 19]
```diff
@@ -307,58 +346,87 @@ For example:
...
+        "created_at": "2014-04-17",
+        "deleted_at": null,
+        "updated_at": "2014-04-18",
+        "revision":2,
+        "parents":
+          [
+            {
+              "name":"Testing Infrastructure",
+              "slugs":["testing", "tests"],
+              "created_at": "2015-04-17",
```
Comment (athai) [inline source/draft_api.rst]: More year inconsistencies (against revision 2).

File: source/draft_api.rst [hunk 20]
```diff
@@ -307,58 +346,87 @@ For example:
...
+        "created_at": "2014-04-17",
+        "deleted_at": null,
+        "updated_at": "2014-04-23",
+        "revision":2,
+        "parents":
+          [
+            {
+              "name":"Testing Infrastructure",
+              "slugs":["testing", "tests"],
+              "created_at": "2015-04-17",
```
Comment (athai) [inline source/draft_api.rst]: Here as well.

File: source/draft_errors.rst [hunk 21]
```diff
@@ -209,35 +198,35 @@ another user.
...
 
 -------------------
 
-10. Request failure
+9. Request Failure
 -------------------
 
 This error is returned when a GET, POST, DELETE, etc. request is sent to an
-object and is rejected. It's used mainly in the instances when a user tries to
+object and is rejected. It"s used mainly in the instances when a user tries to
```
Comment (athai) [inline source/draft_errors.rst]: Why did you change the apostrophe to a quotation?

File: source/draft_errors.rst [hunk 22]
```diff
@@ -41,12 +41,14 @@ Slug) which does not match an object in the database.
 
 ---------------
 
-2. Server error
+2. Server Error
 ---------------
 
-A generic catch-all for when there is a server error outside of the client's
+A generic catch-all for when there is a server error outside of the client"s
```
Comment (athai) [inline source/draft_errors.rst]: Quotation mark -> apostrophe

**Comment by athai** [comment]
I think it's good now, unless you care about spacing inconsistencies in the code-blocks.

**Comment by ElijahCaine** [comment]
@LyonesGamer This still looks good to me, but can you squash the commits before we do any merging? It makes rebasing with other branches a lot easier.

File: source/draft_api.rst [hunk 23]
```diff
@@ -681,25 +729,14 @@ DELETE Endpoints
...
-a 404 Object Not Found error (see error docs) .
+The object will be soft-deleted; that is, the object will still exist in the
+database, but will be unable to be accessed via a normal GET request, returning
+a 404.
 
-If an object is deleted and the ``?include_deleted=true`` parameter is passed
-it will appear in the response in addition to any other objects matching the
-desired query (e.g,. time entries matching a date range).
+Activities and projects cannot be deleted if they are referenced by a current
+(i.e. not deleted or updated) time. This will return a Request Failure.
```
Comment (Kennric) [inline source/draft_api.rst]: What if they are referenced by a deleted time?

**Comment by Kennric** [comment]
Looking pretty good, aside from some ambiguity about deleting things associated with deleted times. I think even deleted times will need to have a reference to what project they were for, so perhaps we should not delete activities or projects if -any- time is associated.

File: source/draft_api.rst [hunk 24]
```diff
@@ -656,50 +716,37 @@ The following content is checked by the API for validity:
...
+the object from the results, and requests for single objects (e.g.
+``GET /times/:uuid``) will return a 404. The parameter ``?include_deleted``
+circumvents this requirement and allows deleted objects to be returned as well.
 
-If the object does not exist or has ``deleted_at`` set to a non-null datetime
-and the ``?include_deleted=true`` parameter is not passed, the API will return
-a 404 Object Not Found error (see error docs) .
+An object's deleted status is indicated by setting its ``deleted_at`` to the time of
+deletion; if the value is null, the object is not deleted. Only the most recent revision
+is set. In addition, activities and projects have their ``slugs`` set to null, in order
```
Comment (athai) [inline source/draft_api.rst]: Projectslugs are hard-deleted. Is that equivalent to an activityslug being set to null?

Comment (MorganEPatch) [inline source/draft_api.rst]: Yeah, because how you get rid of the slugs depends on your implementation. In SQL, we keep them in a separate table and DELETE them on this delete; in something like Mongo, you'd just set them to null. So this is a little more database agnostic.

**Comment by Kennric** [comment]
I think this is good. +1

**Comment by athai** [comment]
+1
